### PR TITLE
Sample vp for Approximations,  deprecate old ADVI

### PR DESCRIPTION
**Pull request by ferrine**
Following #2026

### CODE REVIEW

File: pymc3/variational/advi.py [hunk 1]
```diff
@@ -108,6 +108,9 @@ def advi(vars=None, start=None, model=None, n=5000, accurate_elbo=False,
         and Blei, D. M. (2016). Automatic Differentiation Variational
         Inference. arXiv preprint arXiv:1603.00788.
     """
+    import warnings
+    warnings.warn('Old ADVI interface is deprecated and be removed in future, use pm.ADVI instead',
```
Comment (twiecki) [inline pymc3/variational/advi.py]: `and will be`

**Comment by ferrine** [comment]
Seems like lots of refactoring will be needed for init_nuts

File: pymc3/variational/advi.py [hunk 2]
```diff
@@ -108,6 +108,10 @@ def advi(vars=None, start=None, model=None, n=5000, accurate_elbo=False,
         and Blei, D. M. (2016). Automatic Differentiation Variational
         Inference. arXiv preprint arXiv:1603.00788.
     """
+    import warnings
+    warnings.warn('Old ADVI interface is deprecated and will '
+                  'be removed in future, use pm.ADVI instead',
```
Comment (twiecki) [inline pymc3/variational/advi.py]: Shouldn't we tell people to use `pm.fit()` instead?

File: pymc3/variational/callbacks.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
+import scipy.stats as stats
+import numpy as np
+
+
+class Callback(object):
+    def __call__(self, approx, loss, i):
+        raise NotImplementedError
+
+
+class CheckLossConvergence(Callback):
```
Comment (twiecki) [inline pymc3/variational/callbacks.py]: Very elegant.

We should also add the classic ADVI convergence criterion: https://github.com/pymc-devs/pymc3/blob/master/pymc3/variational/advi.py#L180

**Comment by twiecki** [comment]
@ferrine This starts to look pretty good. Why is init_nuts trickier than what you already have here?

**Comment by fonnesbeck** [comment]
Do we want to do the `sample_vp` renaming here?

File: pymc3/variational/advi.py (hunk 2 above)
Comment (ferrine) [reply pymc3/variational/advi.py]: Maybe

File: pymc3/variational/callbacks.py (hunk 3 above)
Comment (ferrine) [reply pymc3/variational/callbacks.py]: Okay

**Comment by ferrine** [comment]
@fonnesbeck The only option is to rename to `sample_approx` it is long and it's short name is weird

**Comment by ferrine** [comment]
@twiecki why do we need some strange tests for random states?

**Comment by ferrine** [comment]
@twiecki init nuts became a problem because of start point for fine-tuning advi from map. I did not support that case. 
I also did not have cunstructor for inference from it's approx. It is done badly here. But I didn't find more pretty solution. 
So refactoring adds more constructors and start point. I also want to try svgd init

**Comment by fonnesbeck** [comment]
@ferrine `sample_approx` isn't too long. I think being adequately descriptive and intuitive is more important than the number of characters, in general. I thought we wanted to generalize `_vp` away from just variational posteriors?

**Comment by ferrine** [comment]
@fonnesbeck good point. So method of `Approximation` will be renamed to just `sample` in order not to tell obvious things or misinform.

**Comment by fonnesbeck** [comment]
@ferrine that's a good idea, I agree.

**Comment by ferrine** [comment]
Comming back to the tests, they are failing. What for are these tests? https://github.com/pymc-devs/pymc3/blob/master/pymc3/tests/test_sampling.py#L37

**Comment by twiecki** [comment]
That test was added when we fixed a bug where every parallel sampler generated the same samplers because they all used the same seed.

**Comment by ferrine** [comment]
Hmm, interesting

**Comment by ferrine** [comment]
@twiecki I can't find where I use global numpy rng in my code :(

**Comment by twiecki** [comment]
@ferrine Does that test fail locally too?

**Comment by ferrine** [comment]
yes, for sanity check I used the following code:
```python
s1 = np.random.get_state()
approx = pm.fit(
            n=10, method='advi', model=nnet,
            callbacks=[pm.callbacks.CheckLossConvergence(every=1, window_size=2)]
        )
start = approx.sample(draws=4)
cov = approx.cov.eval()
s2 = np.random.get_state()
(s1[0] == s2[0]), (s1[1] == s2[1]).all(), (s1[2] == s2[2]), (s1[3] == s2[3])
--------------
True, True, True, True 
```

**Comment by twiecki** [comment]
Good thing we have that test.

**Comment by ferrine** [comment]
Adding `np.random.uniform()` between s1 and s2 breaks last check. I wonder if test is correct? Should not it check the state of global rng?

**Comment by ferrine** [comment]
I've added numpy random state to watchlist in debugger and found that global random state is changed here: https://github.com/pymc-devs/pymc3/blob/master/pymc3/step_methods/hmc/nuts.py#L161
Is it what is expected or it should use private random state?
I think difference is caused by random state passed to old advi. It gave same result every time. In my code results are not reproducible as I use theano random generator without controlling seed

UPDATE in other words we had an illusion that global random state is untouched

**Comment by ferrine** [comment]
I've added Inference module to docs, how can I check that it will be displayed?

**Comment by fonnesbeck** [comment]
Will there be a progress bar for `sample_vp`?

**Comment by ferrine** [comment]
Do you mean `sample_approx`? There is no need as it is vectorized and done only once for the whole sample size.

**Comment by ferrine** [comment]
What about merge?

File: pymc3/variational/approximations.py [hunk 4]
```diff
@@ -221,31 +232,35 @@ def from_mean_field(cls, mean_field, gpu_compat=False):
...
 
     Parameters
     ----------
     trace : MultiTrace
-    local_rv : dict
-        Experimental for Histogram
-        mapping {model_variable -> local_variable}
+    local_rv : dict[var->tuple]
+        Experimental for Empirical Distribution
+        mapping {model_variable -> local_variable (:math:`\\mu`, math:`\\rho`)}
```
Comment (ferrine) [inline pymc3/variational/approximations.py]: typo

**Comment by fonnesbeck** [comment]
Looks great, thanks!

File: pymc3/sampling.py [hunk 5]
```diff
@@ -566,19 +567,29 @@ def init_nuts(init='ADVI', njobs=1, n_init=500000, model=None,
...
+        cov = approx.cov.eval()
         if njobs == 1:
             start = start[0]
-        cov = np.power(model.dict_to_array(v_params.stds), 2)
     elif init == 'advi_map':
         start = pm.find_MAP()
-        v_params = pm.variational.advi(n=n_init, start=start,
-                                       random_seed=random_seed)
-        cov = np.power(model.dict_to_array(v_params.stds), 2)
+        approx = pm.MeanField(model=model, start=start)
```
Comment (twiecki) [inline pymc3/sampling.py:582]: I'm getting meaningful differences with this change on this model: http://twiecki.github.io/blog/2017/03/14/random-walk-deep-net/

It does not abort when converged and sampling from this initialization is very slow.

Comment (ferrine) [reply pymc3/sampling.py:582]: Do you see any way of improvement?

Comment (ferrine) [reply pymc3/sampling.py:582]: @twiecki I was suggested to try running minimum stopping criteria. Interesting to see how it will affect performance

**Pull request merged**